### DISCUSSION TITLE
Pruduction with custom whitespace is not a branch rule

### DISCUSSION THREAD

**Discussion by NelsonEloi** [foonathan/lexy#64] [Q&A] [answered]
Hi,

Some how I got my grammar in a state I can't figure how to get around it, I known what causes it but can't figure it out.

godbolt link https://godbolt.org/z/b78d3sznG

That doesn't compile, the reason is because of lines 10 and 17
"static constexpr auto whitespace = dsl::ascii::space;"

Commenting those two lines, it will compile, only now I don't have white space skipping

I got into this because I was writing tests for each production and when I added the root one it stopped compiling.

Thank you very much,
Nelson

**Comment by foonathan** [comment, marked as answer]
I've recently changed whitespace skipping. As a consequence, productions that define `whitespace` members can't be branch rules (because they skip whitespace first, which can't be used as a branch condition). However, you only need to add a `whitespace` member to the root production anyway, it is then inherited by all children: https://godbolt.org/z/E7czTsq3v

**Reply by EmmanuelMess** [reply]
> However, you only need to add a whitespace member to the root production anyway, it is then inherited by all children: [godbolt.org/z/E7czTsq3v](https://godbolt.org/z/E7czTsq3v)

That should be added to the compilation error to help

**Reply by foonathan** [reply]
Good idea. I've improved the static_assert message if it requires a branch condition, but haven't added the specific whitespace part - it only confuses people where that isn't the root problem, I think. Instead, I've tried to make it more clear in the tutorial that a whitespace specifier is only necessary in the root production.
